src/Common/precompiled.h

```diff
@@ -176,6 +176,16 @@ inline sint16 _swapEndianS16(sint16 v)
     return (sint16)(((uint16)v >> 8) | ((uint16)v << 8));
 }
 
+inline uint64 _rotl64 (uint64 x, sint8 r)
+{
+    return (x << r) | (x >> (64 - r));
+}
+
+inline uint64 _rotr64 (uint64 x, sint8 r)
+{
+    return (x >> r) | (x << (64 - r));
+}
```

Comment:
C++20 has [`std::rotl`](https://en.cppreference.com/w/cpp/numeric/rotl) and [`std::rotr`](https://en.cppreference.com/w/cpp/numeric/rotr)

Reply:
It causes UB when used with 64 bit values. (Need to investigate further)

Reply:
What UB does it cause? The C++ standard says that the implementation should behave basically as if it was implemented like these two functions, see https://timsong-cpp.github.io/cppwp/n4868/bit.rotate

> Let `r` be `s % N`.
>
> Returns: If `r` is `0`, `x`; if `r` is positive, `(x << r) | (x >> (N - r))`; if `r` is negative, `rotr(x, -r)`.

Reply:
I don't know why but yesterday, when I used it, I got an UB. Might be related to the code I was testing

Reply:
I only wrote these wrappers because _rotl64 and _rotr64 were used in the code and I wanted to get it compiling asap. Since then I've added <bit> to precompiled.h so maybe it's a good idea to just go through and replace all occurrences with the std functions. Thoughts?

Reply:
> I only wrote these wrappers because _rotl64 and _rotr64 were used in the code and I wanted to get it compiling asap. Since then I've added `<bit>` to `precompiled.h` so maybe it's a good idea to just go through and replace all occurrences with the std functions. Thoughts?

Since we are on C++20, it is safe to use rotr or rotl rather than a custom implementation. And also, I found rotr and rotl to be more optimized in godbolt

Reply:
> 

I've gone ahead replaced _rotl64, _rotr64. Slightly different semantics since std::rot on type T also returns T but there's only one place where that matters: https://github.com/tomlally/Cemu/blob/6408cfdd4b391ca47f8296fe002cd1bf8adec37a/src/Cafe/HW/Latte/Core/LatteShader.cpp#L228

Reply:
> 



> > I only wrote these wrappers because _rotl64 and _rotr64 were used in the code and I wanted to get it compiling asap. Since then I've added `<bit>` to `precompiled.h` so maybe it's a good idea to just go through and replace all occurrences with the std functions. Thoughts?
> 
> Since we are on C++20, it is safe to use rotr or rotl rather than a custom implementation. And also, I found rotr and rotl to be more optimized in godbolt

I've gone ahead replaced _rotl64, _rotr64. Slightly different semantics since std::rot on type T also returns T but there's only one place where that matters: https://github.com/tomlally/Cemu/blob/6408cfdd4b391ca47f8296fe002cd1bf8adec37a/src/Cafe/HW/Latte/Core/LatteShader.cpp#L228